lxd/api_cluster.go

```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// Do nothing if the node is already evacuated.
+	if node.State == db.ClusterMemberStateEvacuated {
+		return response.SmartError(fmt.Errorf("Node is already evacuated"))
+	}
+
+	reverter := revert.New()
+	defer reverter.Fail()
+
+	// Ensure node is put into its previous state if anything fails.
+	reverter.Add(func() {
+		d.cluster.Transaction(func(tx *db.ClusterTx) error {
+			tx.UpdateNodeStatus(node.ID, db.ClusterMemberStateCreated)
+
+			return nil
+		})
+	})
+
+	// The instances are retrieved in a separate transaction, after the node is in EVACUATED state.
+	var dbInstances []db.Instance
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// If evacuating, consider only the instances on the node which needs to be evacuated.
+		dbInstances, err = tx.GetInstances(db.InstanceFilter{Node: nodeName})
+		if err != nil {
+			return errors.Wrap(err, "Failed to get instances")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	instances := make([]instance.Instance, len(dbInstances))
+
+	for i, dbInst := range dbInstances {
+		inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+		if err != nil {
+			return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+		}
+
+		instances[i] = inst
+	}
+
+	var targetNodeName string
+	var targetNode db.NodeInfo
+
+	migrateInstances := instance.GetMigratableInstances(d.cluster, instances)
+
+	run := func(op *operations.Operation) error {
+		source, err := cluster.Connect(node.Address, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+		if err != nil {
+			return errors.Wrap(err, "Failed to connect to source")
+		}
+
+		metadata := make(map[string]interface{})
+
+		// Stop all instances if needed
+		for _, inst := range instances {
+			if !inst.IsRunning() {
+				continue
+			}
+
+			metadata["evacuation_progress"] = fmt.Sprintf("Stopping %s", inst.Name())
+			op.UpdateMetadata(metadata)
+
+			err = inst.Stop(false)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to stop instance %q", inst.Name())
+			}
+		}
+
+		// Migrate instances
+		for _, inst := range migrateInstances {
+			metadata["evacuation_progress"] = fmt.Sprintf("Migrating %s", inst.Name())
+			op.UpdateMetadata(metadata)
+
+			inst.VolatileSet(map[string]string{"volatile.evacuate.origin": nodeName})
+
+			err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+				targetNodeName, err = tx.GetNodeWithLeastInstances([]int{node.Architecture}, -1)
+				if err != nil {
+					return err
+				}
+
+				targetNode, err = tx.GetNodeByName(targetNodeName)
+				if err != nil {
+					return err
+				}
+
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+
+			dest, err := cluster.Connect(targetNode.Address, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+			if err != nil {
+				return errors.Wrap(err, "Failed to connect to destination")
+			}
```

Comment:
I don't think you actually ever need to connect to the destination directly.

`source.UpdateInstanceState` should work just fine, LXD will just forward the request internally as needed.

Reply:
Tried that but it hangs forever at `startOp.Wait()` because it's waiting for `chActive`. I didn't investigate any further.

Reply:
Tried passing `false` instead of `true`?

`true` is for notifications within a cluster and so doesn't get you events, in this case, you're doing a real remote operation so you don't want notify mode.

Reply:
If it's set to `false`, I get `Error: Failed to connect to source: Missing event connection with target cluster member`.

Reply:
Hmm, well, that error is a bit of a problem as that also explains the event issue.

Are you getting this because the test is running too early and not all members have established event connections yet?

Reply:
I see there issue here. Basically, the node is connecting to itself (`cluster.Connect(node.Address, ...)` in order to use `MigrateInstance()`. However, there is not listener with the node's own address which causes the failure.

Reply:
Ah, that'd make sense, the whole connecting to itself part is a bit dodgy anyway, it should be able to just make an internal call instead.